File: agent/sdk/python/layerx_sdk/perps.py
"""Typed builders for every LayerX perps module activity.

Each encoder emits the exact bytes the kernel codec (``src/modules/perps``)
accepts and each decoder refuses what the kernel refuses.
"""

from __future__ import annotations

from dataclasses import dataclass, fields
from typing import ClassVar, Literal

PERPS_MODULE_ID = 6
MAX_ORACLE_KEYS = 8
ADL_CAPACITY = 128
BASIS_POINTS_ONE = 10000

TradingPayloadErrorCode = Literal["unknown_activity", "length", "non_canonical", "parameter_bounds", "unsorted_sequence"]


class TradingPayloadError(ValueError):
    def __init__(self, code: TradingPayloadErrorCode, detail: str = "") -> None:
        super().__init__(f"{code}: {detail}" if detail else code)
        self.code = code


_WIDTHS = {"u8": 1, "u32": 4, "u64": 8, "u128": 16}
_ZERO_ID = "00" * 32


def _check_id(value: str, nonzero: bool = True) -> bytes:
    if not isinstance(value, str) or len(value) != 64 or value != value.lower():
        raise TradingPayloadError("non_canonical", "identifier")
    try:
        raw = bytes.fromhex(value)
    except ValueError as error:
        raise TradingPayloadError("non_canonical", "identifier") from error
    if nonzero and raw == bytes(32):
        raise TradingPayloadError("non_canonical", "zero identifier")
    return raw


class Writer:
    def __init__(self) -> None:
        self.out = bytearray()

    def id(self, value: str, nonzero: bool = True) -> Writer:
        self.out += _check_id(value, nonzero)
        return self

    def uint(self, value: int, width: int) -> Writer:
        if not isinstance(value, int) or isinstance(value, bool) or value < 0 or value >= 1 << (8 * width):
            raise TradingPayloadError("non_canonical", "integer width")
        self.out += value.to_bytes(width, "big")
        return self

    def flag(self, value: bool) -> Writer:
        self.out.append(1 if value else 0)
        return self


class Reader:
    def __init__(self, data: bytes) -> None:
        self.data = data
        self.offset = 0

    def take(self, count: int) -> bytes:
        if self.offset + count > len(self.data):
            raise TradingPayloadError("length")
        out = self.data[self.offset : self.offset + count]
        self.offset += count
        return out

    def id(self) -> str:
        return self.take(32).hex()

    def uint(self, width: int) -> int:
        return int.from_bytes(self.take(width), "big")

    def flag(self) -> bool:
        value = self.uint(1)
        if value > 1:
            raise TradingPayloadError("non_canonical", "flag")
        return value == 1


def _one_or_two(value: int, label: str) -> int:
    if value not in (1, 2):
        raise TradingPayloadError("non_canonical", label)
    return value


def _nonzero(*values: int) -> None:
    if any(value == 0 for value in values):
        raise TradingPayloadError("non_canonical", "zero value")


class _Layout:
    """A fixed-width activity whose fields encode in declaration order."""

    ORDINAL: ClassVar[int]
    MODULE: ClassVar[int]
    LAYOUT: ClassVar[tuple[str, ...]]

    @property
    def activity_type(self) -> int:
        return (self.MODULE << 16) | self.ORDINAL

    def check(self) -> None:
        return None

    def encode(self) -> bytes:
        self.check()
        writer = Writer()
        for spec, item in zip(self.LAYOUT, fields(self)):  # type: ignore[arg-type]
            value = getattr(self, item.name)
            if spec == "id":
                writer.id(value)
            elif spec == "flag":
                writer.flag(value)
            elif spec == "choice":
                writer.uint(_one_or_two(value, item.name), 1)
            else:
                writer.uint(value, _WIDTHS[spec])
        return bytes(writer.out)

    @classmethod
    def length(cls) -> int:
        return sum(32 if spec == "id" else 1 if spec in ("flag", "choice") else _WIDTHS[spec] for spec in cls.LAYOUT)

    @classmethod
    def read(cls, reader: Reader) -> _Layout:
        values = []
        for spec, item in zip(cls.LAYOUT, fields(cls)):  # type: ignore[arg-type]
            if spec == "id":
                values.append(reader.id())
            elif spec == "flag":
                values.append(reader.flag())
            elif spec == "choice":
                values.append(_one_or_two(reader.uint(1), item.name))
            else:
                values.append(reader.uint(_WIDTHS[spec]))
        return cls(*values)


class _Perps(_Layout):
    MODULE = PERPS_MODULE_ID


@dataclass(frozen=True)
class PerpsMarketCreate(_Perps):
    market_id: str
    quote_asset: str
    administrator: str
    liquidity_account_id: str
    long_funding_account_id: str
    short_funding_account_id: str
    insurance_account_id: str
    contract_size: int
    tick_size: int
    lot_size: int
    price_scale: int
    initial_margin_ratio_bps: int
    maintenance_margin_ratio_bps: int
    liquidation_fee_bps: int
    liquidator_share_bps: int
    maximum_funding_rate_bps: int
    maximum_deviation_basis_points: int
    funding_interval_ms: int
    maximum_oracle_staleness_ms: int
    minimum_price: int
    maximum_price: int
    permitted_oracle_keys: tuple[str, ...]
    parameter_version: int
    halted: bool

    ORDINAL = 1
    LAYOUT = (
        ("id",) * 7
        + ("u128",) * 4
        + ("u32",) * 6
        + ("u64",) * 2
        + ("u128",) * 2
    )

    def check(self) -> None:
        accounts = [self.liquidity_account_id, self.long_funding_account_id, self.short_funding_account_id, self.insurance_account_id]
        keys = list(self.permitted_oracle_keys)
        ok = (
            self.market_id != _ZERO_ID
            and self.quote_asset != _ZERO_ID
            and self.administrator != _ZERO_ID
            and all(account != _ZERO_ID and account not in accounts[:index] for index, account in enumerate(accounts))
            and self.contract_size != 0
            and self.tick_size != 0
            and self.lot_size != 0
            and self.price_scale != 0
            and self.maintenance_margin_ratio_bps != 0
            and self.maintenance_margin_ratio_bps < self.initial_margin_ratio_bps <= BASIS_POINTS_ONE
            and self.liquidation_fee_bps <= BASIS_POINTS_ONE
            and self.liquidator_share_bps <= BASIS_POINTS_ONE
            and 0 < self.maximum_funding_rate_bps <= BASIS_POINTS_ONE
            and 0 < self.maximum_deviation_basis_points <= BASIS_POINTS_ONE
            and self.funding_interval_ms != 0
            and self.maximum_oracle_staleness_ms != 0
            and 0 < self.minimum_price < self.maximum_price
            and self.parameter_version != 0
            and 0 < len(keys) <= MAX_ORACLE_KEYS
            and all(key != _ZERO_ID for key in keys)
            and all(left < right for left, right in zip(keys, keys[1:]))
        )
        if not ok:
            raise TradingPayloadError("parameter_bounds")

    def encode(self) -> bytes:
        self.check()
        writer = Writer()
        head = [getattr(self, item.name) for item in fields(self)][: len(self.LAYOUT)]
        for spec, value in zip(self.LAYOUT, head):
            if spec == "id":
                writer.id(value)
            else:
                writer.uint(value, _WIDTHS[spec])
        writer.uint(len(self.permitted_oracle_keys), 1)
        for key in self.permitted_oracle_keys:
            writer.id(key)
        writer.out += bytes(32 * (MAX_ORACLE_KEYS - len(self.permitted_oracle_keys)))
        writer.uint(self.parameter_version, 4).flag(self.halted)
        return bytes(writer.out)

    @classmethod
    def length(cls) -> int:
        return 622

    @classmethod
    def read(cls, reader: Reader) -> PerpsMarketCreate:
        head: list[object] = [reader.id() if spec == "id" else reader.uint(_WIDTHS[spec]) for spec in cls.LAYOUT]
        count = reader.uint(1)
        if count > MAX_ORACLE_KEYS:
            raise TradingPayloadError("non_canonical", "oracle key count")
        keys = []
        for index in range(MAX_ORACLE_KEYS):
            key = reader.id()
            if index < count:
                keys.append(key)
            elif key != _ZERO_ID:
                raise TradingPayloadError("non_canonical", "oracle key padding")
        market = cls(*head, tuple(keys), reader.uint(4), reader.flag())  # type: ignore[arg-type]
        market.check()
        return market


@dataclass(frozen=True)
class PerpsMarketHalt(_Perps):
    market_id: str
    halted: bool
    ORDINAL = 2
    LAYOUT = ("id", "flag")


@dataclass(frozen=True)
class PerpsOraclePush(_Perps):
    market_id: str
    observation_sequence: int
    price: int
    observed_at: int
    source_identifier: int
    ORDINAL = 3
    LAYOUT = ("id", "u64", "u128", "u64", "u64")

    def check(self) -> None:
        _nonzero(self.observation_sequence, self.price, self.observed_at, self.source_identifier)


@dataclass(frozen=True)
class PerpsOrderPlace(_Perps):
    market_id: str
    order_id: str
    owner_account_id: str
    side: int
    price: int
    quantity: int
    ORDINAL = 4
    LAYOUT = ("id", "id", "id", "choice", "u128", "u128")

    def check(self) -> None:
        _nonzero(self.price, self.quantity)


@dataclass(frozen=True)
class PerpsOrderCancel(_Perps):
    market_id: str
    order_id: str
    ORDINAL = 5
    LAYOUT = ("id", "id")


@dataclass(frozen=True)
class PerpsPositionOpen(_Perps):
    market_id: str
    position_id: str
    margin_account_id: str
    side: int
    size: int
    entry_notional: int
    margin_amount: int
    ORDINAL = 6
    LAYOUT = ("id", "id", "id", "choice", "u128", "u128", "u128")

    def check(self) -> None:
        _nonzero(self.size, self.margin_amount)
        if self.entry_notional != 0:
            raise TradingPayloadError("non_canonical", "entry notional must be zero")


@dataclass(frozen=True)
class PerpsPositionIncrease(_Perps):
    market_id: str
    position_id: str
    size_delta: int
    notional_delta: int
    margin_amount: int
    ORDINAL = 7
    LAYOUT = ("id", "id", "u128", "u128", "u128")

    def check(self) -> None:
        _nonzero(self.size_delta, self.margin_amount)
        if self.notional_delta != 0:
            raise TradingPayloadError("non_canonical", "notional delta must be zero")


@dataclass(frozen=True)
class PerpsPositionClose(_Perps):
    market_id: str
    position_id: str
    ORDINAL = 8
    LAYOUT = ("id", "id")


@dataclass(frozen=True)
class PerpsFundingTick(_Perps):
    market_id: str
    ORDINAL = 9
    LAYOUT = ("id",)


@dataclass(frozen=True)
class PerpsLiquidate(_Perps):
    market_id: str
    position_id: str
    liquidator_account_id: str
    ORDINAL = 10
    LAYOUT = ("id", "id", "id")


@dataclass(frozen=True)
class PerpsAdl(_Perps):
    market_id: str
    position_ids: tuple[str, ...]
    ORDINAL = 11
    LAYOUT = ()

    def encode(self) -> bytes:
        ids = list(self.position_ids)
        if not 0 < len(ids) <= ADL_CAPACITY:
            raise TradingPayloadError("non_canonical", "position count")
        for position in ids:
            _check_id(position)
        if any(left >= right for left, right in zip(ids, ids[1:])):
            raise TradingPayloadError("unsorted_sequence")
        writer = Writer().id(self.market_id).uint(len(ids), 1)
        for position in ids:
            writer.id(position)
        return bytes(writer.out)


PerpsActivity = (
    PerpsMarketCreate
    | PerpsMarketHalt
    | PerpsOraclePush
    | PerpsOrderPlace
    | PerpsOrderCancel
    | PerpsPositionOpen
    | PerpsPositionIncrease
    | PerpsPositionClose
    | PerpsFundingTick
    | PerpsLiquidate
    | PerpsAdl
)

PERPS_ACTIVITIES: dict[int, type[_Layout]] = {
    kind.ORDINAL: kind
    for kind in (
        PerpsMarketCreate,
        PerpsMarketHalt,
        PerpsOraclePush,
        PerpsOrderPlace,
        PerpsOrderCancel,
        PerpsPositionOpen,
        PerpsPositionIncrease,
        PerpsPositionClose,
        PerpsFundingTick,
        PerpsLiquidate,
        PerpsAdl,
    )
}


def build_perps_activity(activity: PerpsActivity) -> tuple[int, bytes]:
    """The packed activity type and exact payload bytes a perps activity submits."""
    return activity.activity_type, activity.encode()


def decode_trading_activity(
    activities: dict[int, type[_Layout]], module: int, activity_type: int, payload: bytes
) -> _Layout:
    kind = activities.get(activity_type & 0xFFFF) if activity_type >> 16 == module else None
    if kind is None:
        raise TradingPayloadError("unknown_activity", hex(activity_type))
    if kind is PerpsAdl:
        if not 65 <= len(payload) <= 33 + 32 * ADL_CAPACITY:
            raise TradingPayloadError("length")
        reader = Reader(payload)
        market_id = reader.id()
        count = reader.uint(1)
        if count == 0 or len(payload) != 33 + 32 * count or market_id == _ZERO_ID:
            raise TradingPayloadError("non_canonical", "adl shape")
        decoded: _Layout = PerpsAdl(market_id, tuple(reader.id() for _ in range(count)))
        decoded.encode()
        return decoded
    if len(payload) != kind.length():
        raise TradingPayloadError("length")
    decoded = kind.read(Reader(bytes(payload)))
    if decoded.encode() != bytes(payload):
        raise TradingPayloadError("non_canonical")
    return decoded


def decode_perps_activity(activity_type: int, payload: bytes) -> PerpsActivity:
    """Decodes and validates kernel payload bytes of a perps activity."""
    return decode_trading_activity(PERPS_ACTIVITIES, PERPS_MODULE_ID, activity_type, payload)  # type: ignore[return-value]

File: agent/sdk/typescript/src/perps.ts
/**
 * Typed builders for every LayerX perps module activity. Each encoder emits
 * the exact bytes the kernel codec (`src/modules/perps`) accepts and each
 * decoder refuses what the kernel refuses.
 */

const ID = /^[0-9a-f]{64}$/u;
const MAX_ORACLE_KEYS = 8;
const ADL_CAPACITY = 128;
const MARKET_BYTES = 622;
const BASIS_POINTS_ONE = 10000n;

export const PERPS_MODULE_ID = 6;

export const PERPS_ACTIVITY_TYPES = {
  market_create: 0x00060001,
  market_halt: 0x00060002,
  oracle_push: 0x00060003,
  order_place: 0x00060004,
  order_cancel: 0x00060005,
  position_open: 0x00060006,
  position_increase: 0x00060007,
  position_close: 0x00060008,
  funding_tick: 0x00060009,
  liquidate: 0x0006000a,
  adl: 0x0006000b,
} as const;

export type PerpsActivityName = keyof typeof PERPS_ACTIVITY_TYPES;

export type TradingPayloadErrorCode =
  | "unknown_activity"
  | "length"
  | "non_canonical"
  | "parameter_bounds"
  | "unsorted_sequence";

export class TradingPayloadError extends Error {
  readonly code: TradingPayloadErrorCode;

  constructor(code: TradingPayloadErrorCode, detail?: string) {
    super(detail === undefined ? code : `${code}: ${detail}`);
    this.name = "TradingPayloadError";
    this.code = code;
  }
}

/** 1 buys (perps) or bids (spot); 2 sells or asks. */
export type TradeSide = 1 | 2;

/** A 32-byte identifier as 64 lowercase hex digits. */
export type TradingId = string;

export interface PerpsMarket {
  readonly marketId: TradingId;
  readonly quoteAsset: TradingId;
  readonly administrator: TradingId;
  readonly liquidityAccountId: TradingId;
  readonly longFundingAccountId: TradingId;
  readonly shortFundingAccountId: TradingId;
  readonly insuranceAccountId: TradingId;
  readonly contractSize: bigint;
  readonly tickSize: bigint;
  readonly lotSize: bigint;
  readonly priceScale: bigint;
  readonly initialMarginRatioBps: bigint;
  readonly maintenanceMarginRatioBps: bigint;
  readonly liquidationFeeBps: bigint;
  readonly liquidatorShareBps: bigint;
  readonly maximumFundingRateBps: bigint;
  readonly maximumDeviationBasisPoints: bigint;
  readonly fundingIntervalMs: bigint;
  readonly maximumOracleStalenessMs: bigint;
  readonly minimumPrice: bigint;
  readonly maximumPrice: bigint;
  readonly permittedOracleKeys: readonly TradingId[];
  readonly parameterVersion: bigint;
  readonly halted: boolean;
}

export type PerpsActivity =
  | ({ readonly activity: "market_create" } & PerpsMarket)
  | { readonly activity: "market_halt"; readonly marketId: TradingId; readonly halted: boolean }
  | {
      readonly activity: "oracle_push";
      readonly marketId: TradingId;
      readonly observationSequence: bigint;
      readonly price: bigint;
      readonly observedAt: bigint;
      readonly sourceIdentifier: bigint;
    }
  | {
      readonly activity: "order_place";
      readonly marketId: TradingId;
      readonly orderId: TradingId;
      readonly ownerAccountId: TradingId;
      readonly side: TradeSide;
      readonly price: bigint;
      readonly quantity: bigint;
    }
  | { readonly activity: "order_cancel"; readonly marketId: TradingId; readonly orderId: TradingId }
  | {
      readonly activity: "position_open";
      readonly marketId: TradingId;
      readonly positionId: TradingId;
      readonly marginAccountId: TradingId;
      readonly side: TradeSide;
      readonly size: bigint;
      readonly entryNotional: bigint;
      readonly marginAmount: bigint;
    }
  | {
      readonly activity: "position_increase";
      readonly marketId: TradingId;
      readonly positionId: TradingId;
      readonly sizeDelta: bigint;
      readonly notionalDelta: bigint;
      readonly marginAmount: bigint;
    }
  | { readonly activity: "position_close"; readonly marketId: TradingId; readonly positionId: TradingId }
  | { readonly activity: "funding_tick"; readonly marketId: TradingId }
  | {
      readonly activity: "liquidate";
      readonly marketId: TradingId;
      readonly positionId: TradingId;
      readonly liquidatorAccountId: TradingId;
    }
  | { readonly activity: "adl"; readonly marketId: TradingId; readonly positionIds: readonly TradingId[] };

export interface EncodedActivity {
  readonly activityType: number;
  readonly payload: Uint8Array;
}

/** Big-endian writer shared by the perps and spot codecs. */
export class TradingWriter {
  private readonly parts: number[] = [];

  id(value: TradingId, nonzero = true): this {
    if (!ID.test(value)) {
      throw new TradingPayloadError("non_canonical", "identifier");
    }
    if (nonzero && /^0+$/u.test(value)) {
      throw new TradingPayloadError("non_canonical", "zero identifier");
    }
    for (let index = 0; index < 64; index += 2) {
      this.parts.push(Number.parseInt(value.slice(index, index + 2), 16));
    }
    return this;
  }

  uint(value: bigint, bytes: number): this {
    const limit = (1n << BigInt(bytes * 8)) - 1n;
    if (value < 0n || value > limit) {
      throw new TradingPayloadError("non_canonical", "integer width");
    }
    for (let index = bytes - 1; index >= 0; index -= 1) {
      this.parts.push(Number((value >> BigInt(index * 8)) & 0xffn));
    }
    return this;
  }

  byte(value: number): this {
    this.parts.push(value);
    return this;
  }

  zeros(count: number): this {
    for (let index = 0; index < count; index += 1) {
      this.parts.push(0);
    }
    return this;
  }

  bytes(): Uint8Array {
    return Uint8Array.from(this.parts);
  }
}

/** Big-endian reader shared by the perps and spot codecs. */
export class TradingReader {
  private offset = 0;

  constructor(private readonly source: Uint8Array) {}

  private take(count: number): Uint8Array {
    if (this.offset + count > this.source.length) {
      throw new TradingPayloadError("length");
    }
    const out = this.source.subarray(this.offset, this.offset + count);
    this.offset += count;
    return out;
  }

  id(): TradingId {
    return Array.from(this.take(32), (byte) => byte.toString(16).padStart(2, "0")).join("");
  }

  uint(bytes: number): bigint {
    return this.take(bytes).reduce((value, byte) => (value << 8n) | BigInt(byte), 0n);
  }

  byte(): number {
    return this.take(1)[0] ?? 0;
  }

  flag(): boolean {
    const value = this.byte();
    if (value > 1) {
      throw new TradingPayloadError("non_canonical", "flag");
    }
    return value === 1;
  }
}

function nonzero(...values: bigint[]): void {
  if (values.some((value) => value === 0n)) {
    throw new TradingPayloadError("non_canonical", "zero value");
  }
}

function side(value: number): TradeSide {
  if (value !== 1 && value !== 2) {
    throw new TradingPayloadError("non_canonical", "side");
  }
  return value;
}

function isZeroId(value: TradingId): boolean {
  return /^0+$/u.test(value);
}

function validateMarket(market: PerpsMarket): void {
  const accounts = [
    market.liquidityAccountId,
    market.longFundingAccountId,
    market.shortFundingAccountId,
    market.insuranceAccountId,
  ];
  const keys = market.permittedOracleKeys;
  const ok =
    !isZeroId(market.marketId) &&
    !isZeroId(market.quoteAsset) &&
    !isZeroId(market.administrator) &&
    accounts.every((account, index) => !isZeroId(account) && !accounts.slice(0, index).includes(account)) &&
    market.contractSize !== 0n &&
    market.tickSize !== 0n &&
    market.lotSize !== 0n &&
    market.priceScale !== 0n &&
    market.maintenanceMarginRatioBps !== 0n &&
    market.initialMarginRatioBps > market.maintenanceMarginRatioBps &&
    market.initialMarginRatioBps <= BASIS_POINTS_ONE &&
    market.liquidationFeeBps <= BASIS_POINTS_ONE &&
    market.liquidatorShareBps <= BASIS_POINTS_ONE &&
    market.maximumFundingRateBps !== 0n &&
    market.maximumFundingRateBps <= BASIS_POINTS_ONE &&
    market.maximumDeviationBasisPoints !== 0n &&
    market.maximumDeviationBasisPoints <= BASIS_POINTS_ONE &&
    market.fundingIntervalMs !== 0n &&
    market.maximumOracleStalenessMs !== 0n &&
    market.minimumPrice !== 0n &&
    market.minimumPrice < market.maximumPrice &&
    market.parameterVersion !== 0n &&
    keys.length > 0 &&
    keys.length <= MAX_ORACLE_KEYS &&
    keys.every((key, index) => ID.test(key) && !isZeroId(key) && (index === 0 || (keys[index - 1] ?? "") < key));
  if (!ok) {
    throw new TradingPayloadError("parameter_bounds");
  }
}

function encodeMarket(writer: TradingWriter, market: PerpsMarket): void {
  validateMarket(market);
  writer
    .id(market.marketId)
    .id(market.quoteAsset)
    .id(market.administrator)
    .id(market.liquidityAccountId)
    .id(market.longFundingAccountId)
    .id(market.shortFundingAccountId)
    .id(market.insuranceAccountId)
    .uint(market.contractSize, 16)
    .uint(market.tickSize, 16)
    .uint(market.lotSize, 16)
    .uint(market.priceScale, 16)
    .uint(market.initialMarginRatioBps, 4)
    .uint(market.maintenanceMarginRatioBps, 4)
    .uint(market.liquidationFeeBps, 4)
    .uint(market.liquidatorShareBps, 4)
    .uint(market.maximumFundingRateBps, 4)
    .uint(market.maximumDeviationBasisPoints, 4)
    .uint(market.fundingIntervalMs, 8)
    .uint(market.maximumOracleStalenessMs, 8)
    .uint(market.minimumPrice, 16)
    .uint(market.maximumPrice, 16)
    .byte(market.permittedOracleKeys.length);
  for (const key of market.permittedOracleKeys) {
    writer.id(key);
  }
  writer
    .zeros((MAX_ORACLE_KEYS - market.permittedOracleKeys.length) * 32)
    .uint(market.parameterVersion, 4)
    .byte(market.halted ? 1 : 0);
}

function decodeMarket(reader: TradingReader): PerpsActivity {
  const head = {
    marketId: reader.id(),
    quoteAsset: reader.id(),
    administrator: reader.id(),
    liquidityAccountId: reader.id(),
    longFundingAccountId: reader.id(),
    shortFundingAccountId: reader.id(),
    insuranceAccountId: reader.id(),
    contractSize: reader.uint(16),
    tickSize: reader.uint(16),
    lotSize: reader.uint(16),
    priceScale: reader.uint(16),
    initialMarginRatioBps: reader.uint(4),
    maintenanceMarginRatioBps: reader.uint(4),
    liquidationFeeBps: reader.uint(4),
    liquidatorShareBps: reader.uint(4),
    maximumFundingRateBps: reader.uint(4),
    maximumDeviationBasisPoints: reader.uint(4),
    fundingIntervalMs: reader.uint(8),
    maximumOracleStalenessMs: reader.uint(8),
    minimumPrice: reader.uint(16),
    maximumPrice: reader.uint(16),
  };
  const count = reader.byte();
  if (count > MAX_ORACLE_KEYS) {
    throw new TradingPayloadError("non_canonical", "oracle key count");
  }
  const permittedOracleKeys: TradingId[] = [];
  for (let index = 0; index < MAX_ORACLE_KEYS; index += 1) {
    const key = reader.id();
    if (index < count) {
      permittedOracleKeys.push(key);
    } else if (!isZeroId(key)) {
      throw new TradingPayloadError("non_canonical", "oracle key padding");
    }
  }
  const market = {
    ...head,
    permittedOracleKeys,
    parameterVersion: reader.uint(4),
    halted: reader.flag(),
  };
  validateMarket(market);
  return { activity: "market_create", ...market };
}

function encodeAdl(writer: TradingWriter, marketId: TradingId, positionIds: readonly TradingId[]): void {
  if (positionIds.length === 0 || positionIds.length > ADL_CAPACITY) {
    throw new TradingPayloadError("non_canonical", "position count");
  }
  if (positionIds.some((id) => !ID.test(id) || isZeroId(id))) {
    throw new TradingPayloadError("non_canonical", "position identifier");
  }
  if (positionIds.some((id, index) => index > 0 && (positionIds[index - 1] ?? "") >= id)) {
    throw new TradingPayloadError("unsorted_sequence");
  }
  writer.id(marketId).byte(positionIds.length);
  for (const id of positionIds) {
    writer.id(id);
  }
}

/** Returns the packed activity type of a perps activity. */
export function perpsActivityType(activity: PerpsActivity): number {
  return PERPS_ACTIVITY_TYPES[activity.activity];
}

/** Encodes the exact kernel payload bytes of a perps activity. */
export function encodePerpsActivity(activity: PerpsActivity): Uint8Array {
  const writer = new TradingWriter();
  switch (activity.activity) {
    case "market_create":
      encodeMarket(writer, activity);
      break;
    case "market_halt":
      writer.id(activity.marketId).byte(activity.halted ? 1 : 0);
      break;
    case "oracle_push":
      nonzero(activity.observationSequence, activity.price, activity.observedAt, activity.sourceIdentifier);
      writer
        .id(activity.marketId)
        .uint(activity.observationSequence, 8)
        .uint(activity.price, 16)
        .uint(activity.observedAt, 8)
        .uint(activity.sourceIdentifier, 8);
      break;
    case "order_place":
      nonzero(activity.price, activity.quantity);
      writer
        .id(activity.marketId)
        .id(activity.orderId)
        .id(activity.ownerAccountId)
        .byte(side(activity.side))
        .uint(activity.price, 16)
        .uint(activity.quantity, 16);
      break;
    case "order_cancel":
      writer.id(activity.marketId).id(activity.orderId);
      break;
    case "position_open":
      nonzero(activity.size, activity.marginAmount);
      if (activity.entryNotional !== 0n) {
        throw new TradingPayloadError("non_canonical", "entry notional must be zero");
      }
      writer
        .id(activity.marketId)
        .id(activity.positionId)
        .id(activity.marginAccountId)
        .byte(side(activity.side))
        .uint(activity.size, 16)
        .uint(activity.entryNotional, 16)
        .uint(activity.marginAmount, 16);
      break;
    case "position_increase":
      nonzero(activity.sizeDelta, activity.marginAmount);
      if (activity.notionalDelta !== 0n) {
        throw new TradingPayloadError("non_canonical", "notional delta must be zero");
      }
      writer
        .id(activity.marketId)
        .id(activity.positionId)
        .uint(activity.sizeDelta, 16)
        .uint(activity.notionalDelta, 16)
        .uint(activity.marginAmount, 16);
      break;
    case "position_close":
      writer.id(activity.marketId).id(activity.positionId);
      break;
    case "funding_tick":
      writer.id(activity.marketId);
      break;
    case "liquidate":
      writer.id(activity.marketId).id(activity.positionId).id(activity.liquidatorAccountId);
      break;
    case "adl":
      encodeAdl(writer, activity.marketId, activity.positionIds);
      break;
  }
  return writer.bytes();
}

/** Builds the activity type and payload a perps activity submits. */
export function buildPerpsActivity(activity: PerpsActivity): EncodedActivity {
  return { activityType: perpsActivityType(activity), payload: encodePerpsActivity(activity) };
}

const PERPS_LENGTHS: Readonly<Record<number, number>> = {
  1: MARKET_BYTES,
  2: 33,
  3: 72,
  4: 129,
  5: 64,
  6: 145,
  7: 112,
  8: 64,
  9: 32,
  10: 96,
};

function sameBytes(left: Uint8Array, right: Uint8Array): boolean {
  return left.length === right.length && left.every((byte, index) => byte === right[index]);
}

/** Decodes and validates kernel payload bytes of a perps activity. */
export function decodePerpsActivity(activityType: number, payload: Uint8Array): PerpsActivity {
  if (activityType >>> 16 !== PERPS_MODULE_ID) {
    throw new TradingPayloadError("unknown_activity", activityType.toString(16));
  }
  const ordinal = activityType & 0xffff;
  if (ordinal === 11) {
    if (payload.length < 65 || payload.length > 33 + 32 * ADL_CAPACITY) {
      throw new TradingPayloadError("length");
    }
    const reader = new TradingReader(payload);
    const marketId = reader.id();
    const count = reader.byte();
    if (count === 0 || payload.length !== 33 + 32 * count || isZeroId(marketId)) {
      throw new TradingPayloadError("non_canonical", "adl shape");
    }
    const positionIds = Array.from({ length: count }, () => reader.id());
    const decoded: PerpsActivity = { activity: "adl", marketId, positionIds };
    encodePerpsActivity(decoded);
    return decoded;
  }
  const expected = PERPS_LENGTHS[ordinal];
  if (expected === undefined) {
    throw new TradingPayloadError("unknown_activity", activityType.toString(16));
  }
  if (payload.length !== expected) {
    throw new TradingPayloadError("length");
  }
  const reader = new TradingReader(payload);
  let decoded: PerpsActivity;
  switch (ordinal) {
    case 1:
      decoded = decodeMarket(reader);
      break;
    case 2:
      decoded = { activity: "market_halt", marketId: reader.id(), halted: reader.flag() };
      break;
    case 3:
      decoded = {
        activity: "oracle_push",
        marketId: reader.id(),
        observationSequence: reader.uint(8),
        price: reader.uint(16),
        observedAt: reader.uint(8),
        sourceIdentifier: reader.uint(8),
      };
      break;
    case 4:
      decoded = {
        activity: "order_place",
        marketId: reader.id(),
        orderId: reader.id(),
        ownerAccountId: reader.id(),
        side: side(reader.byte()),
        price: reader.uint(16),
        quantity: reader.uint(16),
      };
      break;
    case 5:
      decoded = { activity: "order_cancel", marketId: reader.id(), orderId: reader.id() };
      break;
    case 6:
      decoded = {
        activity: "position_open",
        marketId: reader.id(),
        positionId: reader.id(),
        marginAccountId: reader.id(),
        side: side(reader.byte()),
        size: reader.uint(16),
        entryNotional: reader.uint(16),
        marginAmount: reader.uint(16),
      };
      break;
    case 7:
      decoded = {
        activity: "position_increase",
        marketId: reader.id(),
        positionId: reader.id(),
        sizeDelta: reader.uint(16),
        notionalDelta: reader.uint(16),
        marginAmount: reader.uint(16),
      };
      break;
    case 8:
      decoded = { activity: "position_close", marketId: reader.id(), positionId: reader.id() };
      break;
    case 9:
      decoded = { activity: "funding_tick", marketId: reader.id() };
      break;
    default:
      decoded = {
        activity: "liquidate",
        marketId: reader.id(),
        positionId: reader.id(),
        liquidatorAccountId: reader.id(),
      };
  }
  if (!sameBytes(encodePerpsActivity(decoded), payload)) {
    throw new TradingPayloadError("non_canonical");
  }
  return decoded;
}
